crates/sui-graphql-rpc/src/types/epoch.rs

```diff
@@ -87,6 +95,37 @@ impl Epoch {
         )))
     }
 
+    /// The total number of transaction blocks in this epoch so far.
+    async fn total_transactions(&self, ctx: &Context<'_>) -> Result<Option<BigInt>> {
+        match self.stored.epoch_total_transactions {
+            Some(total) => Ok(Some(BigInt::from(total as u64))),
+            None => {
+                // If the epoch has not ended yet, we calculate the total transactions by
+                // subtracting the last epoch's network total transactions from the last checkpoint's.
+                let last_checkpoint_network_total =
+                    Checkpoint::query(ctx.data_unchecked(), CheckpointId::default())
+                        .await
+                        .extend()?
+                        .map(|c| c.network_total_transactions_impl());
+                let last_epoch_last_checkpoint_id =
+                    CheckpointId::by_seq_num((self.stored.first_checkpoint_id - 1) as u64);
+                let last_epoch_network_total =
+                    Checkpoint::query(ctx.data_unchecked(), last_epoch_last_checkpoint_id)
+                        .await
+                        .extend()?
+                        .map(|c| c.network_total_transactions_impl());
+                match (last_checkpoint_network_total, last_epoch_network_total) {
+                    (Some(last_checkpoint_network_total), Some(last_epoch_network_total)) => {
+                        Ok(Some(BigInt::from(
+                            last_checkpoint_network_total - last_epoch_network_total,
+                        )))
+                    }
+                    _ => Ok(None),
+                }
+            }
+        }
+    }
```

Comment:
I would be a little surprised if this was the issue in the test, given how linear it is, but this implementation is susceptible to races -- we are issuing separate queries for the various checkpoints involved, and not bounding them to make sure they are all within the epoch (imagine that this query is issued right as the epoch changes, then your query for the latest checkpoint may actually fetch a latest checkpoint from an even later epoch).

The way to fix this today would be to issue a query for the latest checkpoint within the epoch, but the real fix depends on @wlmyng's consistent read work.  #15962 adds a `checkpoint_viewed_at` field to the `Epoch` type.  If `epoch_total_transactions` is not set, then it's that checkpoint that you need to read the transactions from.

I would suggest returning `None` for this field on the latest epoch in this PR, and then building on top of @wlmyng's stack with the proper fix afterwards.